src/fcntl.rs

```diff
@@ -188,19 +188,21 @@ fn wrap_readlink_result(v: &mut Vec<u8>, res: ssize_t) -> Result<OsString> {
 }
 
 pub fn readlink<'a, P: ?Sized + NixPath>(path: &P) -> Result<OsString> {
-    let mut v = vec![0u8; libc::PATH_MAX as usize];
+    let len = libc::PATH_MAX as usize;
+    let mut v = Vec::with_capacity(len);
     let res = path.with_nix_path(|cstr| {
-        unsafe { libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.len() as size_t) }
+        unsafe { libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, len as size_t) }
```

Comment:
You seem to be unaware of the `capacity` method.  You should write it like this:
```rust
let mut v = Vec::with_capacity(libc::PATH_MAX as usize);
let res = path.with_nix_path(|cstr| {
    unsafe { libc::readlink(cstr.as_ptr(), v.as_mut_ptr() as *mut c_char, v.capacity() as size_t) }
```
```